src/core/polyhedral/codegen_llvm.cc

```diff
@@ -355,16 +355,34 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
   }
 };
 
+template <typename CI, typename CO, typename Pred>
+bool all(const CI& inputs, const CO& outputs, Pred p) {
+  for (const auto& v : inputs) {
+    if (not p(v)) {
+      return false;
+    }
+  }
+  for (const auto& v : outputs) {
+    if (not p(v)) {
+      return false;
+    }
+  }
+  return true;
+}
+
 class LLVMCodegen {
-  void collectTensor(const Halide::OutputImageParam& t) {
+  llvm::Type* convertTensorToType(const Halide::OutputImageParam& t) {
```

Comment:
why return a raw pointer? how do we ensure it doesn't leak later?

Reply:
In LLVM almost everything is passed around as a pointer. The associated `LLVMContext` keeps track of everything.